pkg/telemetry/client.go

```diff
@@ -0,0 +1,274 @@
+	if metrics == nil {
+		return errors.New("post metrics data is required for update")
+	}
+	tc.currentOperationMetrics.ExitStatus = metrics.ExitCode
+	tc.currentOperationMetrics.EndTime = time.Now()
+	return nil
+}
+
+func (tc *telemetryClient) SaveMetrics() error {
+	// If cli command fail cobra validation, the PersistentPreRunE() wouldn't be invoked where initialization is done
+	// so, it is safe to ignore the metrics for user errors(like typos) at least to an extent where cobra can validate.
+	if tc.currentOperationMetrics.StartTime.IsZero() {
+		return nil
+	}
+
+	err := tc.metricsDB.CreateSchema()
+	if err != nil {
+		return errors.Wrap(err, "unable to create the telemetry schema")
+	}
+
+	return tc.metricsDB.SaveOperationMetric(tc.currentOperationMetrics)
+}
+
+// SendMetrics sends the local stored metrics to super collider
+// TODO: to be implemented
+func (tc *telemetryClient) SendMetrics() error {
+	return nil
+}
+
+func (tc *telemetryClient) isCoreCommand(cmdPath string) bool {
+	if cmdPath != "" {
+		cmds := strings.Split(cmdPath, " ")
+		if len(cmds) >= 2 && tc.coreCommandMap[cmds[1]] {
+			return true
+		}
+	}
+
+	return false
+}
+func (tc *telemetryClient) updateMetricsForCoreCommand(cmd *cobra.Command, args []string, cliID string) error {
+	tc.currentOperationMetrics.CliID = cliID
+	tc.currentOperationMetrics.CliVersion = buildinfo.Version
+	tc.currentOperationMetrics.StartTime = time.Now()
+	tc.currentOperationMetrics.CommandName = strings.Join(strings.Split(cmd.CommandPath(), " ")[1:], " ")
+
+	// CLI recommendation is to have a single argument for a command
+	if len(args) != 0 {
+		tc.currentOperationMetrics.NameArg = hashString(args[0])
+	}
+
+	flagMap := make(map[string]string)
+	hashRequired := isHashRequiredForCmdFlags(cmd.CommandPath())
+	cmd.Flags().Visit(func(flag *pflag.Flag) {
+		// capture the boolean and empty flag values as is
+		if !hashRequired || flag.Value.String() == "" || flag.Value.Type() == "bool" {
+			flagMap[flag.Name] = flag.Value.String()
+		} else {
+			flagMap[flag.Name] = hashString(flag.Value.String())
+		}
+	})
+	if len(flagMap) != 0 {
+		jsonString, _ := json.Marshal(flagMap)
+		tc.currentOperationMetrics.Flags = string(jsonString)
+	}
+
+	return nil
+}
+
+func (tc *telemetryClient) updateMetricsForPlugin(cmd *cobra.Command, args []string, cliID string) error {
+	tc.currentOperationMetrics.CliID = cliID
+	tc.currentOperationMetrics.CliVersion = buildinfo.Version
+	tc.currentOperationMetrics.StartTime = time.Now()
+
+	flagNames := TraverseFlagNames(args)
+	if len(flagNames) > 0 {
+		tc.currentOperationMetrics.Flags = flagNamesToJSONString(flagNames)
+	}
+
+	plugin := tc.pluginInfoFromCommand(cmd)
+	if plugin != nil {
+		tc.currentOperationMetrics.PluginName = plugin.Name
+		tc.currentOperationMetrics.PluginVersion = plugin.Version
+		tc.currentOperationMetrics.Target = string(plugin.Target)
+		tc.currentOperationMetrics.Endpoint = getEndpointSHA(plugin)
```

Comment:
What does the `Endpoint` mean here?

Reply:
`endpoint` is the endpoint that the command is directed at.
It could be kubernetes cluster or SaaS endpoint.